File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseKGroup(ListNode head, int k) {
        ListNode prevTail = null;
        ListNode nextHead = null;

        ListNode slow = head;
        ListNode fast = head;

        for(int i = 0;i < k;i++)
        {
            if(fast != null)
            {
                fast = fast.next;
            }
            else
            {
                return head;
            }
        }

        nextHead = fast;

        while(true)
        {
            ListNode temp = slow;
            ListNode next = null;
            ListNode prev = null;
            while(temp != fast)
            {
                next = temp.next;
                temp.next = prev;
                prev = temp;
                temp = next;
            }

            if(prevTail != null)
            {
                prevTail.next = prev;
            }
            else
            {
                head = prev;
            }
            prevTail = slow;
            slow.next = nextHead;
            slow = slow.next;
            for(int i = 0;i < k;i++)
            {
                if(fast != null)
                {
                    fast = fast.next;
                }
                else
                {
                    return head;
                }
            }

            nextHead = fast;
        }

        return head;
    }
}

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if (head == nullptr || head->next == nullptr)
            return head;

        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* cur = dummy;

        ListNode* leftEnd = head;
        ListNode* rightEnd = head;

        for (int i = 1; i < k; i++) {
            if (rightEnd->next != nullptr) {
                rightEnd = rightEnd->next;
            } else {
                return head;
            }
        }
        ListNode* afterRight = rightEnd->next;
        rightEnd->next = nullptr;

        cur->next = reverse(leftEnd);
        leftEnd->next = reverseKGroup(afterRight, k);
        return dummy->next;
    }

private:
    ListNode* reverse(ListNode* head) {
        if (head == nullptr || head->next == nullptr)
            return head;

        ListNode* newHead = reverse(head->next);
        head->next->next = head;
        head->next = nullptr;

        return newHead;
    }
};